File: lib/qdrant.ts
import { QdrantClient } from '@qdrant/js-client-rest';
import * as dotenv from 'dotenv';
import * as path from 'path';
import { getQueryEmbedding } from './embeddings';

// Load backend/.env as well as root .env
dotenv.config({ path: path.resolve(process.cwd(), 'backend/.env') });
dotenv.config();

export const COLLECTION_NAME = 'vtu_study_companion';

/**
 * Authoritative 14 First-Year CSE / AIML subjects
 * Physics Cycle:
 * 1. Mathematics
 * 2. Physics
 * 3. Communication English
 * 4. Kannada Kali / Manasu
 * 5. A Scientific Approach to Health
 * 6. Principles of Programming Using C
 * 7. ESC
 * 8. PLC
 * 
 * Chemistry Cycle:
 * 9. Chemistry
 * 10. Professional Writing Skills in English
 * 11. Constitution of India
 * 12. Design Thinking
 * 13. Computer-Aided Engineering Drawing
 * 14. ETC
 */
export const AUTHORITATIVE_FIRST_YEAR_SUBJECTS: string[] = [
  "Mathematics",
  "Physics",
  "Communication English",
  "Kannada Kali / Manasu",
  "A Scientific Approach to Health",
  "Principles of Programming Using C",
  "ESC",
  "PLC",
  "Chemistry",
  "Professional Writing Skills in English",
  "Constitution of India",
  "Design Thinking",
  "Computer-Aided Engineering Drawing",
  "ETC"
];

// Set of explicitly excluded second-year or extraneous subjects
export const EXCLUDED_SUBJECTS = new Set<string>([
  "AI and ML",
  "Analysis and Design of Algorithms",
  "Computer Networks",
  "Data Science",
  "Database Management Systems",
  "Microcontrollers",
  "Operating Systems",
  "Software Engineering"
]);

// Map raw Qdrant payload subject strings to canonical first-year names
export const RAW_TO_CANONICAL_SUBJECT_MAP: Record<string, string> = {
  "CAED": "Computer-Aided Engineering Drawing",
  "Computer-Aided Engineering Drawing": "Computer-Aided Engineering Drawing",
  "Chemistry": "Chemistry",
  "Communication English": "Communication English",
  "Constitution of India": "Constitution of India",
  "Design Thinking": "Design Thinking",
  "ESC": "ESC",
  "ETC": "ETC",
  "Kannada Kali Manasu": "Kannada Kali / Manasu",
  "Kannada Kali / Manasu": "Kannada Kali / Manasu",
  "Mathematics ChemistryCycle": "Mathematics",
  "Mathematics PhysicsCycle": "Mathematics",
  "Mathematics": "Mathematics",
  "Physics": "Physics",
  "PLC": "PLC",
  "Principles of Programming C": "Principles of Programming Using C",
  "Principles of Programming Using C": "Principles of Programming Using C",
  "Professional Writing English": "Professional Writing Skills in English",
  "Professional Writing Skills in English": "Professional Writing Skills in English",
  "Scientific Approach to Health": "A Scientific Approach to Health",
  "A Scientific Approach to Health": "A Scientific Approach to Health"
};

// Map canonical first-year subjects to all matching raw subject values in Qdrant payloads
export const CANONICAL_TO_RAW_SUBJECT_MAP: Record<string, string[]> = {
  "Mathematics": ["Mathematics", "Mathematics ChemistryCycle", "Mathematics PhysicsCycle"],
  "Physics": ["Physics"],
  "Communication English": ["Communication English"],
  "Kannada Kali / Manasu": ["Kannada Kali / Manasu", "Kannada Kali Manasu"],
  "A Scientific Approach to Health": ["A Scientific Approach to Health", "Scientific Approach to Health"],
  "Principles of Programming Using C": ["Principles of Programming Using C", "Principles of Programming C"],
  "ESC": ["ESC"],
  "PLC": ["PLC"],
  "Chemistry": ["Chemistry"],
  "Professional Writing Skills in English": ["Professional Writing Skills in English", "Professional Writing English"],
  "Constitution of India": ["Constitution of India"],
  "Design Thinking": ["Design Thinking"],
  "Computer-Aided Engineering Drawing": ["Computer-Aided Engineering Drawing", "CAED"],
  "ETC": ["ETC"]
};

export function getRawSubjectFilterValues(subject: string): string[] {
  if (!subject) return [];
  const trimmed = subject.trim();
  if (CANONICAL_TO_RAW_SUBJECT_MAP[trimmed]) {
    return CANONICAL_TO_RAW_SUBJECT_MAP[trimmed];
  }
  const canonical = RAW_TO_CANONICAL_SUBJECT_MAP[trimmed];
  if (canonical && CANONICAL_TO_RAW_SUBJECT_MAP[canonical]) {
    return CANONICAL_TO_RAW_SUBJECT_MAP[canonical];
  }
  return [trimmed];
}

let clientInstance: QdrantClient | null = null;

/**
 * Retrieves a singleton instance of the Qdrant Client.
 */
export function getQdrantClient(): QdrantClient {
  if (!clientInstance) {
    const url = process.env.QDRANT_URL;
    const apiKey = process.env.QDRANT_API_KEY;

    if (!url || !apiKey) {
      console.warn('[WARNING] QDRANT_URL or QDRANT_API_KEY is missing.');
    }
    
    clientInstance = new QdrantClient({
      url: url || 'http://localhost:6333',
      apiKey: apiKey || '',
      checkCompatibility: false,
    });
  }
  return clientInstance;
}

/**
 * Fetch collection statistics and extract the 14 authoritative first-year subjects.
 */
export async function getCollectionStats() {
  const client = getQdrantClient();
  try {
    const collectionInfo = await client.getCollection(COLLECTION_NAME);
    const total_chunks = collectionInfo.points_count || 0;
    
    return {
      total_chunks,
      subjects: [...AUTHORITATIVE_FIRST_YEAR_SUBJECTS],
      doc_types: {}
    };
  } catch (error) {
    console.error('[ERROR] Qdrant stats failed:', error);
    return {
      total_chunks: 0,
      subjects: [...AUTHORITATIVE_FIRST_YEAR_SUBJECTS],
      doc_types: {}
    };
  }
}

/**
 * Fetch all chunks (used by Predict endpoint)
 */
export async function getAllChunks(where?: any): Promise<Array<{text: string, metadata: any}>> {
  const client = getQdrantClient();
  try {
    const out: Array<{text: string, metadata: any}> = [];
    let offset: any = undefined;
    
    const subjectEq = where?.subject?.$eq;
    let filter: any = undefined;

    if (subjectEq) {
      const rawValues = getRawSubjectFilterValues(subjectEq);
      filter = {
        must: [
          rawValues.length === 1
            ? { key: 'subject', match: { value: rawValues[0] } }
            : { key: 'subject', match: { any: rawValues } }
        ]
      };
    }

    do {
      const response = await client.scroll(COLLECTION_NAME, {
        filter: filter,
        limit: 1000,
        offset: offset,
        with_payload: true,
        with_vector: false
      });
      
      for (const p of response.points) {
        const payload = (p.payload || {}) as Record<string, any>;
        const rawSubj = payload.subject;
        if (rawSubj && EXCLUDED_SUBJECTS.has(rawSubj)) {
          continue;
        }

        const text = (payload.text as string) || '';
        const metadata = { ...payload };
        delete metadata.text;

        if (metadata.subject && RAW_TO_CANONICAL_SUBJECT_MAP[metadata.subject]) {
          metadata.subject = RAW_TO_CANONICAL_SUBJECT_MAP[metadata.subject];
        }
        
        out.push({ text, metadata });
      }
      
      offset = response.next_page_offset;
    } while (offset !== null && offset !== undefined);
    
    return out;
  } catch (error) {
    console.error('[ERROR] Qdrant scroll failed:', error);
    return [];
  }
}

export interface SearchResult {
  text: string;
  subject: string;
  unit: string;
  doc_type: string;
  filename: string;
  score: number;
}

/**
 * Perform a semantic vector search
 */
export async function semanticSearch(
  query: string,
  n_results = 5,
  subject?: string,
  unit?: string,
  doc_type?: string
): Promise<SearchResult[]> {
  const client = getQdrantClient();
  
  try {
    const qEmb = await getQueryEmbedding(query);
    
    const mustConditions: any[] = [];
    if (subject) {
      const rawValues = getRawSubjectFilterValues(subject);
      if (rawValues.length === 1) {
        mustConditions.push({ key: 'subject', match: { value: rawValues[0] } });
      } else if (rawValues.length > 1) {
        mustConditions.push({ key: 'subject', match: { any: rawValues } });
      }
    }
    if (unit) mustConditions.push({ key: 'unit', match: { value: unit } });
    if (doc_type) mustConditions.push({ key: 'doc_type', match: { value: doc_type } });
    
    const filter = mustConditions.length > 0 ? { must: mustConditions } : undefined;
    
    const results = await client.query(COLLECTION_NAME, {
      query: qEmb,
      filter: filter,
      limit: n_results,
      with_payload: true,
    });
    
    return results.points
      .filter((r: any) => {
        const rawSubj = r.payload?.subject;
        return !(rawSubj && EXCLUDED_SUBJECTS.has(rawSubj));
      })
      .map((r: any) => {
        const rawSubj = (r.payload?.subject as string) || '';
        const canonicalSubj = RAW_TO_CANONICAL_SUBJECT_MAP[rawSubj] || rawSubj;
        return {
          text: (r.payload?.text as string) || '',
          subject: canonicalSubj,
          unit: (r.payload?.unit as string) || '',
          doc_type: (r.payload?.doc_type as string) || '',
          filename: (r.payload?.filename as string) || '',
          score: r.score,
        };
      });
  } catch (error) {
    console.error('[ERROR] Qdrant search failed:', error);
    return [];
  }
}

/**
 * Query chunks for legacy wrapper
 */
export async function queryChunks(q: string, n = 5, where?: any): Promise<SearchResult[]> {
  const subject = where?.subject?.$eq;
  return semanticSearch(q, n, subject);
}

File: backend/lib/clients/qdrant.py
import os
import uuid
import hashlib
from typing import List, Dict, Optional, Any
from qdrant_client import QdrantClient
from qdrant_client.models import PointStruct, VectorParams, Distance, Filter, MatchValue, FieldCondition

# PUBLIC EXPORTS
COLLECTION_NAME = os.getenv("QDRANT_COLLECTION_NAME", "vtu_study_companion")
EMBEDDING_MODEL = os.getenv("EMBEDDING_MODEL", "BAAI/bge-small-en-v1.5")

AUTHORITATIVE_FIRST_YEAR_SUBJECTS = [
    "Mathematics",
    "Physics",
    "Communication English",
    "Kannada Kali / Manasu",
    "A Scientific Approach to Health",
    "Principles of Programming Using C",
    "ESC",
    "PLC",
    "Chemistry",
    "Professional Writing Skills in English",
    "Constitution of India",
    "Design Thinking",
    "Computer-Aided Engineering Drawing",
    "ETC"
]

EXCLUDED_SUBJECTS = {
    "AI and ML",
    "Analysis and Design of Algorithms",
    "Computer Networks",
    "Data Science",
    "Database Management Systems",
    "Microcontrollers",
    "Operating Systems",
    "Software Engineering"
}

RAW_TO_CANONICAL_SUBJECT_MAP = {
    "CAED": "Computer-Aided Engineering Drawing",
    "Computer-Aided Engineering Drawing": "Computer-Aided Engineering Drawing",
    "Chemistry": "Chemistry",
    "Communication English": "Communication English",
    "Constitution of India": "Constitution of India",
    "Design Thinking": "Design Thinking",
    "ESC": "ESC",
    "ETC": "ETC",
    "Kannada Kali Manasu": "Kannada Kali / Manasu",
    "Kannada Kali / Manasu": "Kannada Kali / Manasu",
    "Mathematics ChemistryCycle": "Mathematics",
    "Mathematics PhysicsCycle": "Mathematics",
    "Mathematics": "Mathematics",
    "Physics": "Physics",
    "PLC": "PLC",
    "Principles of Programming C": "Principles of Programming Using C",
    "Principles of Programming Using C": "Principles of Programming Using C",
    "Professional Writing English": "Professional Writing Skills in English",
    "Professional Writing Skills in English": "Professional Writing Skills in English",
    "Scientific Approach to Health": "A Scientific Approach to Health",
    "A Scientific Approach to Health": "A Scientific Approach to Health"
}

CANONICAL_TO_RAW_SUBJECT_MAP = {
    "Mathematics": ["Mathematics", "Mathematics ChemistryCycle", "Mathematics PhysicsCycle"],
    "Physics": ["Physics"],
    "Communication English": ["Communication English"],
    "Kannada Kali / Manasu": ["Kannada Kali / Manasu", "Kannada Kali Manasu"],
    "A Scientific Approach to Health": ["A Scientific Approach to Health", "Scientific Approach to Health"],
    "Principles of Programming Using C": ["Principles of Programming Using C", "Principles of Programming C"],
    "ESC": ["ESC"],
    "PLC": ["PLC"],
    "Chemistry": ["Chemistry"],
    "Professional Writing Skills in English": ["Professional Writing Skills in English", "Professional Writing English"],
    "Constitution of India": ["Constitution of India"],
    "Design Thinking": ["Design Thinking"],
    "Computer-Aided Engineering Drawing": ["Computer-Aided Engineering Drawing", "CAED"],
    "ETC": ["ETC"]
}

def get_raw_subject_filter_values(subject: str) -> List[str]:
    if not subject:
        return []
    s = subject.strip()
    if s in CANONICAL_TO_RAW_SUBJECT_MAP:
        return CANONICAL_TO_RAW_SUBJECT_MAP[s]
    canonical = RAW_TO_CANONICAL_SUBJECT_MAP.get(s)
    if canonical and canonical in CANONICAL_TO_RAW_SUBJECT_MAP:
        return CANONICAL_TO_RAW_SUBJECT_MAP[canonical]
    return [s]

_client: Optional[QdrantClient] = None
_embedding_model = None


def _get_embedding_fn():
    global _embedding_model
    if _embedding_model is None:
        from fastembed import TextEmbedding
        _embedding_model = TextEmbedding(model_name=EMBEDDING_MODEL)
    return _embedding_model


def get_qdrant_client() -> Optional[QdrantClient]:
    """
    Returns a Qdrant client. Priority:
    1. Qdrant Cloud (QDRANT_URL + QDRANT_API_KEY)
    2. Local embedded Qdrant (persistent path ./qdrant_data)
    3. In-memory Qdrant (fallback, data lost on restart)
    """
    global _client
    if _client is not None:
        return _client

    url = os.getenv("QDRANT_URL", "").strip()
    key = os.getenv("QDRANT_API_KEY", "").strip()

    if url and key:
        try:
            candidate = QdrantClient(url=url, api_key=key, timeout=15, check_compatibility=False)
            # Verify the connection actually works by listing collections
            candidate.get_collections()
            _client = candidate
            print(f"[OK] Qdrant Cloud connected: {url}")
            return _client
        except Exception as e:
            print(f"[WARN] Qdrant Cloud failed ({e}), falling back to local embedded mode.")

    # Fallback: local embedded Qdrant with a persistent path
    data_path = os.getenv("QDRANT_DATA_PATH", "./qdrant_data")
    try:
        os.makedirs(data_path, exist_ok=True)
        _client = QdrantClient(path=data_path)
        print(f"[OK] Qdrant embedded mode: {data_path}")
        return _client
    except Exception as e:
        print(f"[WARN] Qdrant embedded failed ({e}), using in-memory mode.")

    # Last resort: in-memory (data lost on restart)
    _client = QdrantClient(":memory:")
    print("[WARN] Qdrant running in-memory — data will not persist across restarts.")
    return _client


def get_or_create_collection() -> Optional[QdrantClient]:
    client = get_qdrant_client()
    if not client:
        return None
    try:
        existing = [c.name for c in client.get_collections().collections]
        if COLLECTION_NAME not in existing:
            client.create_collection(
                collection_name=COLLECTION_NAME,
                vectors_config=VectorParams(size=384, distance=Distance.COSINE),
            )
            print(f"[OK] Created Qdrant collection: {COLLECTION_NAME}")
        return client
    except Exception as e:
        print(f"[ERROR] Qdrant collection setup failed: {e}")
        return None


def _generate_deterministic_uuid(filename: str, chunk_index: int) -> str:
    hash_hex = hashlib.md5(f"{filename}_{chunk_index}".encode("utf-8")).hexdigest()
    return f"{hash_hex[:8]}-{hash_hex[8:12]}-{hash_hex[12:16]}-{hash_hex[16:20]}-{hash_hex[20:]}"


def add_chunks(chunks: List[str], metadatas: List[Dict]) -> int:
    client = get_or_create_collection()
    if not client:
        return 0
    embed_fn = _get_embedding_fn()
    points = []
    for t, m, emb in zip(chunks, metadatas, embed_fn.embed(chunks)):
        filename = m.get("filename", "unknown")
        chunk_idx = m.get("chunk_index", 0)
        point_id = _generate_deterministic_uuid(filename, chunk_idx)
        points.append(PointStruct(id=point_id, vector=list(emb), payload={"text": t, **m}))
    if points:
        client.upsert(collection_name=COLLECTION_NAME, points=points)
    return len(points)


def semantic_search(
    query: str,
    n_results: int = 5,
    subject: Optional[str] = None,
    unit: Optional[str] = None,
    doc_type: Optional[str] = None,
) -> List[Dict]:
    client = get_qdrant_client()
    if not client:
        return []
    try:
        embed_fn = _get_embedding_fn()
        q_emb = [float(x) for x in list(embed_fn.embed([query]))[0]]
        filters = []
        if subject:
            raw_subs = get_raw_subject_filter_values(subject)
            if len(raw_subs) == 1:
                filters.append(FieldCondition(key="subject", match=MatchValue(value=raw_subs[0])))
            elif len(raw_subs) > 1:
                from qdrant_client.models import MatchAny
                filters.append(FieldCondition(key="subject", match=MatchAny(any=raw_subs)))
        if unit:
            filters.append(FieldCondition(key="unit", match=MatchValue(value=unit)))
        if doc_type:
            filters.append(FieldCondition(key="doc_type", match=MatchValue(value=doc_type)))

        results = client.query_points(
            collection_name=COLLECTION_NAME,
            query=q_emb,
            query_filter=Filter(must=filters) if filters else None,
            limit=n_results,
            with_payload=True,
        )
        out = []
        for r in results.points:
            raw_subj = r.payload.get("subject", "")
            if raw_subj in EXCLUDED_SUBJECTS:
                continue
            canonical_subj = RAW_TO_CANONICAL_SUBJECT_MAP.get(raw_subj, raw_subj)
            out.append({
                "text": r.payload.get("text", ""),
                "subject": canonical_subj,
                "unit": r.payload.get("unit", ""),
                "doc_type": r.payload.get("doc_type", ""),
                "filename": r.payload.get("filename", ""),
                "score": r.score,
            })
        return out
    except Exception as e:
        print(f"[ERROR] Qdrant search failed: {e}")
        return []


def get_all_chunks(where: Optional[Dict] = None) -> List[Dict]:
    client = get_qdrant_client()
    if not client:
        return []
    try:
        filters = []
        if where and "subject" in where:
            subj_val = where["subject"].get("$eq") if isinstance(where["subject"], dict) else where["subject"]
            if subj_val:
                raw_subs = get_raw_subject_filter_values(subj_val)
                if len(raw_subs) == 1:
                    filters.append(FieldCondition(key="subject", match=MatchValue(value=raw_subs[0])))
                elif len(raw_subs) > 1:
                    from qdrant_client.models import MatchAny
                    filters.append(FieldCondition(key="subject", match=MatchAny(any=raw_subs)))

        out, offset = [], None
        while True:
            pts, offset = client.scroll(
                collection_name=COLLECTION_NAME,
                scroll_filter=Filter(must=filters) if filters else None,
                limit=1000,
                offset=offset,
                with_payload=True,
            )
            for p in pts:
                raw_subj = p.payload.get("subject", "")
                if raw_subj in EXCLUDED_SUBJECTS:
                    continue
                canonical_subj = RAW_TO_CANONICAL_SUBJECT_MAP.get(raw_subj, raw_subj)
                meta = {k: v for k, v in p.payload.items() if k != "text"}
                meta["subject"] = canonical_subj
                out.append({
                    "text": p.payload.get("text", ""),
                    "metadata": meta,
                })
            if not offset:
                break
        return out
    except Exception as e:
        print(f"[ERROR] Qdrant scroll failed: {e}")
        return []


def get_collection_stats() -> Dict:
    client = get_qdrant_client()
    if not client:
        return {"total_chunks": 0, "subjects": list(AUTHORITATIVE_FIRST_YEAR_SUBJECTS), "doc_types": {}}
    try:
        total = client.count(collection_name=COLLECTION_NAME).count
        return {"total_chunks": total, "subjects": list(AUTHORITATIVE_FIRST_YEAR_SUBJECTS), "doc_types": {}}
    except Exception as e:
        print(f"[ERROR] Qdrant stats failed: {e}")
        return {"total_chunks": 0, "subjects": list(AUTHORITATIVE_FIRST_YEAR_SUBJECTS), "doc_types": {}}


def delete_subject(subject: str) -> int:
    client = get_qdrant_client()
    if not client:
        return 0
    try:
        raw_subs = get_raw_subject_filter_values(subject)
        for s in raw_subs:
            client.delete(
                collection_name=COLLECTION_NAME,
                points_selector=Filter(must=[FieldCondition(key="subject", match=MatchValue(value=s))]),
            )
        return 1
    except Exception:
        return 0


def query_chunks(q: str, n: int = 5, where: Optional[Dict] = None) -> List[Dict]:
    subject = where.get("subject", {}).get("$eq") if where else None
    return semantic_search(q, n, subject=subject)


def list_collections() -> List[str]:
    c = get_qdrant_client()
    return [x.name for x in c.get_collections().collections] if c else []


def delete_collection(name: str) -> bool:
    c = get_qdrant_client()
    return c.delete_collection(collection_name=name) if c else False

